### PR TITLE
[dag] support crash recovery for order rule

### PR DESCRIPTION
**Pull request by zekun000**
Add trait/impl for buffer manager adapter and implement the persist/recovery flow.

### CODE REVIEW

**Review by sasha8 — COMMENTED** [review]
I did not find any issue here, but I have a related thought.

In a possible scenario, I might RB my node and crash even before I collected the signatures.
The others will not accept any other node from me in this round because they persist what they signed in the DB. Therefore, after recovery, I might be unable to RB a node unless I persist the node before I RB it. But even then, I have to make sure I broadcast again my last node upon recovery. Otherwise, if all crash at the same point we may stall.

File: consensus/consensus-types/src/block_data.rs [hunk 1]
```diff
@@ -211,6 +211,33 @@ impl BlockData {
...
+                VoteData::new(BlockInfo::empty(), BlockInfo::empty()),
+                LedgerInfoWithSignatures::new(
+                    LedgerInfo::new(BlockInfo::empty(), HashValue::zero()),
+                    AggregateSignature::empty(),
+                ),
+            ),
+            block_type: BlockType::Proposal {
+                payload,
+                author,
+                failed_authors,
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/consensus-types/src/block_data.rs:236]: we need this to recover the anchor election, for those committed (gc'ed) nodes, we need to read this from storage.

File: consensus/consensus-types/src/block_data.rs [hunk 2]
```diff
@@ -211,6 +211,33 @@ impl BlockData {
...
+    ) -> Self {
+        Self {
+            epoch,
+            round,
+            timestamp_usecs,
+            quorum_cert: QuorumCert::new(
+                VoteData::new(BlockInfo::empty(), BlockInfo::empty()),
+                LedgerInfoWithSignatures::new(
+                    LedgerInfo::new(BlockInfo::empty(), HashValue::zero()),
+                    AggregateSignature::empty(),
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/consensus-types/src/block_data.rs:230]: Yes, I'll explore how to get rid of this, make buffer manager take an enum

File: consensus/consensus-types/src/common.rs [hunk 3]
```diff
@@ -90,6 +103,17 @@ impl ProofWithData {
             status: Arc::new(Mutex::new(None)),
         }
     }
+
+    pub fn extend(&mut self, other: ProofWithData) {
+        let other_data_status = other.status.lock().as_mut().unwrap().take();
+        self.proofs.extend(other.proofs);
+        let mut status = self.status.lock();
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/consensus-types/src/common.rs:110]: I think this is just for api interior mutability not concurrency, but I just copied the code

File: consensus/src/dag/adapter.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        let block = ExecutedBlock::new(
+            Block::new_proposal_for_dag(epoch, round, timestamp, payload, author, failed_author)?,
+            StateComputeResult::new_dummy(),
+        );
+        let block_info = block.block_info();
+        let storage = self.storage.clone();
+        Ok(self.executor_channel.unbounded_send(OrderedBlocks {
+            ordered_blocks: vec![block],
+            ordered_proof: LedgerInfoWithSignatures::new(
+                LedgerInfo::new(block_info, HashValue::zero()),
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/dag/adapter.rs:68]: I don't think this part actually works, it's placeholder for now

**Review by ibalajiarun — APPROVED** [review]

**Review by sasha8 — APPROVED** [review]

**Pull request merged**